### PR TITLE
Refactor and update the token standard

### PR DESCRIPTION
**Pull request by Maximkaaa**
* Removed duplicated code between `transfer` and `tranferIncludeFee`
* Removed most 'expect's
* Renamed module erc20_transactions to icrc1_transfer
* Added some unit tests
* Added `FeeRatio` type to simplify check for correct fee ratios
* Add test cases for zero amount transactions, token amount overflows,
  self transfers etc.
* Fix a bug when self-transfer to default subaccount is allowed.
* Fix a bug when deduplication check permits a duplicated transaction if
  its `created_at_date` ts differs from IC time by some margin.
* Move claim tests from icrc1 module (they are not related to icrc).
* Changes the dispaly method to make it clear that the displayed value is
an account not a principal, and also adds the subaccount to the
displayed value. Also adds some unit tests for account.
* Removed unused error variants
* Update token standard

### CODE REVIEW

File: src/token/api/src/account.rs [hunk 1]
```diff
@@ -1,21 +1,24 @@
...
 use crate::error::TxError;
 
 pub static DEFAULT_SUBACCOUNT: Subaccount = [0u8; 32];
 
-#[derive(Debug, Clone, CandidType, Deserialize, Copy, PartialEq, Eq, Serialize)]
+#[derive(Debug, Clone, CandidType, Deserialize, Copy, PartialEq, Eq)]
 pub struct Account {
     pub owner: Principal,
-    pub subaccount: Option<Subaccount>,
+    pub subaccount: Subaccount,
```
Comment (blutooth) [inline src/token/api/src/account.rs:13]: why is subaccount not an opt as in the did file for the ICRC-1 spec? https://github.com/dfinity/ICRC-1/blob/main/ICRC-1.did

File: src/token/api/src/canister.rs [hunk 2]
```diff
@@ -348,126 +315,489 @@ pub trait TokenCanisterAPI: Canister + Sized {
...
+            AccountIdentifier::new(alice().into(), Some(SubaccountIdentifier(alice_sub)));
+        let bob_aid = AccountIdentifier::new(bob().into(), Some(SubaccountIdentifier(bob_sub)));
+
+        let (ctx, canister) = test_context();
+        ctx.update_caller(john());
+
+        assert!(canister
+            .mint(
+                canister.owner(),
+                Some(alice_aid.to_address()),
```
Comment (blutooth) [inline src/token/api/src/canister.rs:556]: so, this would just be for default accounts?

File: src/factory/src/api.rs [hunk 3]
```diff
@@ -68,8 +71,9 @@ impl TokenFactoryCanister {
 
     #[init]
     pub fn init(&self, controller: Principal, ledger_principal: Option<Principal>) {
-        let ledger = ledger_principal
-            .unwrap_or_else(|| Principal::from_text(DEFAULT_LEDGER_PRINCIPAL).unwrap());
+        let ledger = ledger_principal.unwrap_or_else(|| {
+            Principal::from_text(DEFAULT_LEDGER_PRINCIPAL).expect("const value conversion")
```
Comment (hagsteel) [inline src/factory/src/api.rs:75]: Can we use the actual `const` function here instead? `Principal::from_slice` I think it is

File: src/token/api/src/account.rs [hunk 4]
```diff
@@ -27,7 +30,16 @@ impl From<Principal> for Account {
...
 impl Display for Account {
     fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
-        write!(f, "{}", self.owner)
+        if self.subaccount == DEFAULT_SUBACCOUNT {
+            write!(f, "Account({})", self.owner)
+        } else {
+            write!(f, "Account({}, ", self.owner)?;
+            for b in self.subaccount {
+                write!(f, "{b:02X}")?;
+            }
```
Comment (hagsteel) [inline src/token/api/src/account.rs:39]: If we write the comma before the sub account it won't look so naff if there is no sub accounts
```suggestion
            write!(f, "Account({}", self.owner)?;
            for b in self.subaccount {
                write!(f, ", {b:02X}")?;
            }
```

File: src/token/api/src/canister.rs [hunk 5]
```diff
@@ -103,189 +89,170 @@ pub trait TokenCanisterAPI: Canister + Sized {
...
         Ok(())
     }
 
-    /********************** TRANSFERS ***********************/
-    #[cfg_attr(feature = "transfer", update(trait = true))]
-    fn icrc1_transfer(&self, transfer: TransferArgs) -> Result<u128, TransferError> {
-        let recipient = Account::new(transfer.to.owner, transfer.to.subaccount);
-
-        let account = CheckedAccount::with_recipient(recipient, transfer.from_subaccount)?;
+    /********************** BALANCCES INFO ***********************/
```
Comment (hagsteel) [inline src/token/api/src/canister.rs:150]:
```suggestion
    /********************** BALANCES INFO ***********************/
```

File: src/token/api/src/canister/is20_transactions.rs [hunk 6]
```diff
@@ -1,78 +1,306 @@
...
+    amount: Tokens128,
+    fee: Tokens128,
+    fee_to: Account,
+    auction_fee_ratio: FeeRatio,
+) -> Result<(), TxError> {
+    if amount.is_zero() {
+        return Err(TxError::AmountTooSmall);
+    }
+
+    // We use `updaets` structure because sometimes from or to can be equal to fee_to or even to
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:74]:
```suggestion
    // We use `updates` structure because sometimes from or to can be equal to fee_to or even to
```

File: src/token/api/src/canister/is20_transactions.rs [hunk 7]
```diff
@@ -1,78 +1,306 @@
...
+    // auction_account, so we must take carefull approach.
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:75]:
```suggestion
    // auction_account, so we must take a careful approach.
```

File: src/token/api/src/canister/is20_transactions.rs [hunk 8]
```diff
@@ -1,78 +1,306 @@
...
+    let mut updates = Balances::default();
+    updates.set_balance(from, balances.balance_of(from));
+    updates.set_balance(to, balances.balance_of(to));
+    updates.set_balance(fee_to, balances.balance_of(fee_to));
+    updates.set_balance(auction_account(), balances.balance_of(auction_account()));
+
+    let from_balance = updates.balance_of(from);
+
+    // If `amount + fee` overflows max `Tokens128` value, the balance cannot be larger then this
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:84]:
```suggestion
    // If `amount + fee` overflows max `Tokens128` value, the balance cannot be larger than this
```

File: src/token/api/src/canister/is20_transactions.rs [hunk 9]
```diff
@@ -1,78 +1,306 @@
...
+    caller: Principal,
+    to: Account,
+    amount: Tokens128,
+) -> TxReceipt {
+    let total_supply = state.balances.total_supply();
+    if (total_supply + amount).is_none() {
+        // If we allow to mint more then Tokens128::MAX then simplie operation such as getting
+        // total supply or token stats will panic, So we add this check to prevent this.
+        return Err(TxError::AmountOverflow);
     }
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:177]: The comment is redundant

```suggestion
(total_supply + amount).ok_or(TxError::AmountOverflow)?;
```

File: src/token/api/src/canister/is20_transactions.rs [hunk 10]
```diff
@@ -1,78 +1,306 @@
...
+
+pub fn burn(
+    state: &mut CanisterState,
+    caller: Principal,
+    from: Account,
+    amount: Tokens128,
+) -> TxReceipt {
+    let balance = state.balances.balance_of(from);
+
+    if !amount.is_zero() && balance == Tokens128::ZERO {
```
Comment (hagsteel) [inline src/token/api/src/canister/is20_transactions.rs:227]: Pick either, as long as it's consistent

File: src/token/api/src/error.rs [hunk 11]
```diff
@@ -32,14 +26,13 @@ pub enum TxError {
     AmountOverflow,
     #[error("Account is not found")]
     AccountNotFound,
-    #[error("{message}")]
-    GenericError { message: String },
-    #[error("Claim not Allowed")]
-    ClaimNotAllowed,
-    #[error("Temporary unavailable")]
-    TemporaryUnavailable,
+    #[error("No claimable tokens are on the requested subaccount")]
```
Comment (hagsteel) [inline src/token/api/src/error.rs:29]:
```suggestion
    #[error("no tokens to claim on the requested subaccount.")]
```
Also consistency: all error messages should be lower cased

File: src/token/api/src/state.rs [hunk 12]
```diff
@@ -2,24 +2,25 @@ use std::collections::HashMap;
...
 #[derive(Debug, Default, CandidType, Deserialize, IcStorage)]
 pub struct CanisterState {
-    pub bidding_state: BiddingState,
     pub balances: Balances,
-    pub auction_history: AuctionHistory,
     pub stats: StatsData,
     pub ledger: Ledger,
+
+    // We leave this field here not to introduce new version of the state.
+    #[deprecated(note = "Claims are now stored in owner's subaccounts.")]
```
Comment (hagsteel) [inline src/token/api/src/state.rs:23]:
```suggestion
    // We leave this field here to not introduce a new version of the state.
    #[deprecated(note = "claims are now stored in the owner's subaccounts.")]
```

**Review by hagsteel — CHANGES_REQUESTED** [review]
Looks good. Just a few comments

File: src/token/impl/src/canister.rs [hunk 13]
```diff
@@ -96,21 +151,19 @@ mod test {
...
+        canister.state().borrow_mut().stats.name = "To Kill a Mockingbird".to_string();
         // ... write the state to stable storage
-        canister.__pre_upgrade_inst();
+        canister.__pre_upgrade();
 
         // Update the value without writing it to stable storage
-        let mut state = canister.state.borrow_mut();
-        state.bidding_state.fee_ratio = 0.0;
-        drop(state);
+        canister.state().borrow_mut().stats.name = "David Coperfield".to_string();
```
Comment (blutooth) [inline src/token/impl/src/canister.rs:159]: Copperfield

File: src/token/api/src/canister.rs (hunk 2 above)
Comment (Maximkaaa) [reply src/token/api/src/canister.rs:556]: It works just like before - subaccounts can be used to specify any AccountIdentifier, but `claim` method transfers tokens to the default account of the claiming user.

File: src/token/api/src/account.rs (hunk 4 above)
Comment (Maximkaaa) [reply src/token/api/src/account.rs:39]: If there's no subaccount, another brunch is executed. Adding `,` before each byte in subaccount wouldn't look nice....

File: src/token/api/src/canister/is20_transactions.rs (hunk 9 above)
Comment (Maximkaaa) [reply src/token/api/src/canister/is20_transactions.rs:177]: Let me disagree on this one. Because I had to think whether I should add this check and what are the implications of not having it...

File: src/token/api/src/canister/is20_transactions.rs (hunk 10 above)
Comment (Maximkaaa) [reply src/token/api/src/canister/is20_transactions.rs:227]: I just wanted to show my eloquence :)

File: src/token/api/src/canister/is20_transactions.rs (hunk 9 above)
Comment (hagsteel) [reply src/token/api/src/canister/is20_transactions.rs:177]: How can `Tokens128` be more than `Tokens128::MAX`?
I don't think I know which part you are disagreeing on, could you elaborate? (I'm all for disagreement)

Comment (Maximkaaa) [reply src/token/api/src/canister/is20_transactions.rs:177]: Tokens128 cannot, but the total amount is combined of many users' balances, so if we don't check it, total balance can go over Tokens128::MAX.

**Review by blutooth — APPROVED** [review]

**Review by blutooth — APPROVED** [review]

**Comment by blutooth** [comment]
PR looks good, before we merge can we  update the tags on the AMM to a tag of this branch with a PR on the AMM to check it build and the tests pass correctly.

**Comment by Maximkaaa** [comment]
I've created a draft PR for AMM: https://github.com/infinity-swap/v1-amm/pull/224. All tests pass locally except one, but it cannot be related to IS20 update. Waiting for CI build.

**Comment by Maximkaaa** [comment]
(My AMM PR doesn't pass on CI because of the broken CI script when using branch dependency....)
